### PR TITLE
apply API changes to BigDecimal that described in issue 1469

### PR DESCRIPTION
**Pull request by xsalefter**
As described, this work to undo all `intValue()` and `longValue()` that used to make sure the code is not broken too much before changes in API (described in #1469 ) implemented.

### CODE REVIEW

File: invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java [hunk 1]
```diff
@@ -227,18 +225,18 @@ protected UsageInArrearAggregate getToBeBilledUsageDetails(final LocalDate start
             // We generate an entry if we consumed anything on this tier or if this is the first tier to also support $0 Usage item
             if (hasPreviousUsage) {
                 // FIXME-1469 change to correct BigDecimal implementation
-                final Long previousUsageQuantity = tierNum <= lastPreviousUsageTier ? previousUsage.get(tierNum - 1).getQuantity().longValue() : 0;
+                final BigDecimal previousUsageQuantity = tierNum <= lastPreviousUsageTier ? previousUsage.get(tierNum - 1).getQuantity() : BigDecimal.ZERO;
```
Comment (sbrossie) [inline invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java:228]: The current code is converting the  `tieredBlock#Max` (double) into a `long`, later requiring the conversion `previousUsageQuantity#longValue`. I think the options are to
1. Also update our [catalog api](https://github.com/killbill/killbill-api/blob/master/src/main/java/org/killbill/billing/catalog/api/TieredBlock.java#L24) to specify a `BigDecimal`, OR
2. Make a conversion   `tieredBlock#Max` (double) into a `BigDecimal` avoiding the conversion you need later `previousUsageQuantity#longValue`

File: invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java [hunk 2]
```diff
@@ -248,9 +246,8 @@ protected UsageInArrearAggregate getToBeBilledUsageDetails(final LocalDate start
         return toBeBilledDetails;
     }
 
-    UsageConsumableInArrearTierUnitAggregate computeToBeBilledConsumableInArrearWith_TOP_TIER(final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final Long units) throws CatalogApiException {
-
-        Long remainingUnits = units;
+    UsageConsumableInArrearTierUnitAggregate computeToBeBilledConsumableInArrearWith_TOP_TIER(final List<TieredBlock> tieredBlocks, final List<UsageConsumableInArrearTierUnitAggregate> previousUsage, final BigDecimal units) throws CatalogApiException {
+        Long remainingUnits = units.longValue();
```
Comment (sbrossie) [inline invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalConsumableUsageInArrear.java:250]: We don't want any of these `longValue` conversion, the code should only manipulate `BigDecimal`, so I think something like that (untested):

```
           // If we don't change our catalog api, if not we can get rid of this conversion as well
            final BigDecimal tierBlockMax = new BigDecimal(tieredBlock.getMax());
            final BigDecimal tierBlockSize = new BigDecimal(tieredBlock.getSize());

            final BigDecimal[] divRem =  remainingUnits.divideAndRemainder(tierBlockSize);
            final BigDecimal tmp =  (divRem[1].compareTo(BigDecimal.ZERO) == 0) ? divRem[0] : divRem[0].add(BigDecimal.ONE);
            if (tmp.compareTo(tierBlockMax) > 0) {
                remainingUnits = remainingUnits.subtract(tierBlockMax.multiply(tierBlockSize));
            } else {
                targetBlock = tieredBlock;
                targetTierNum = tierNum;
                break;
            }

```

**Review by sbrossie — CHANGES_REQUESTED** [review]
It's looking pretty good, but there are a few things to address.

Also, can you resync your changes with the latest `work-for-1533-0.23.x`

**Comment by pierre** [comment]
@xsalefter Could you pull and merge `work-for-release-0.23.x-usage` into your branch?
Specifically, https://github.com/killbill/killbill/commit/18d679067fae9fcd2eac08934637609faa50acf5 should hopefully fix CI by pulling a `killbill-api` artifact with your API changes.

**Review by sbrossie — COMMENTED** [review]
It's looking good but with regards to this [comment](https://github.com/killbill/killbill/pull/1579#discussion_r797086749), we will also make the api changes to change these fields into `BigDecimal`:
* TieredBlock#getMax
* TieredBlockPriceOverride#getMax
* Block#getSize
* Block#getMinTopUpCredit
* BlockPriceOverride#getSize
* Limit#getMax
* Limit#getMin

I suggest you simply address @pierre 's [comment](https://github.com/killbill/killbill/pull/1579#issuecomment-1027943989) and then we'll merge this PR *as-is*.

Then you can open a new PR and make the changes above for addressing the catalog change apis.

**Comment by xsalefter** [comment]
@pierre done. 
@sbrossie not sure why this PR still have asking about "requested changes" from you. Seems like it refers to yesterday's PR.

**Comment by pierre** [comment]
> @pierre done.

Thanks @xsalefter!

I've confirmed that with my latest GitHub Actions changes, the API `SNAPSHOT` jar with your changes was correctly downloaded, and the tests ran:

```
2022-02-03T03:48:50.2912980Z [INFO] Downloading from sonatype-snapshots: https://oss.sonatype.org/content/repositories/snapshots/org/kill-bill/billing/killbill-api/0.54.0-0c2ca16-SNAPSHOT/killbill-api-0.54.0-0c2ca16-20220202.132348-1.jar
```

**Pull request merged**